crates/bevy_ecs/src/query/state.rs

```diff
@@ -160,51 +163,95 @@ impl<D: QueryData, F: QueryFilter> QueryState<D, F> {
     }
 
     /// Creates a new [`QueryState`] from a given [`World`] and inherits the result of `world.id()`.
-    pub fn new(world: &mut World) -> Self {
-        let mut state = Self::new_uninitialized(world);
+    /// This does not check access of nested queries, so [`Self::update_external_component_access`]
+    /// must be called before querying using this state or returning it to safe code.
+    ///
+    /// # Safety
+    ///
+    /// Before calling any other methods on the returned `QueryState`,
+    /// [`Self::update_external_component_access`] must be called.
+    pub unsafe fn new_unchecked(world: &mut World) -> Self {
+        let fetch_state = D::init_state(world);
+        let filter_state = F::init_state(world);
+        // SAFETY: Caller ensures `update_external_component_access` is called
+        let mut state =
+            unsafe { Self::from_states_uninitialized(world, fetch_state, filter_state) };
         state.update_archetypes(world);
         state
     }
 
+    /// Collects the access from this query and any nested queries
+    /// and panics
+    pub fn update_external_component_access(
+        &self,
+        system_name: Option<&str>,
+        component_access_set: &mut FilteredAccessSet,
+        world: UnsafeWorldCell,
+    ) {
+        let conflicts = component_access_set.get_conflicts_single(&self.component_access);
+        if !conflicts.is_empty() {
+            let mut accesses = conflicts.format_conflict_list(world);
+            // Access list may be empty (if access to all components requested)
+            if !accesses.is_empty() {
+                accesses.push(' ');
+            }
+            let type_name = DebugName::type_name::<Query<D, F>>();
+            let type_name = type_name.shortname();
+            let system = system_name
+                .map(|name| format!(" in system {name}"))
+                .unwrap_or_default();
+            panic!("error[B0001]: {type_name}{system} accesses component(s) {accesses}in a way that conflicts with a previous system parameter. Consider using `Without<T>` to create disjoint Queries or merging conflicting Queries into a `ParamSet`. See: https://bevy.org/learn/errors/b0001",);
+        }
+
+        component_access_set.add(self.component_access.clone());
+        D::update_external_component_access(
```

Comment:
Wouldn't we want to collect the access from nested queries before checking if they cause conflicts with other SystemParams? Or would the conflict become apparent later?

It would be nice to have at least one implementation of a QueryData that implements `update_external_component_access` so we can add unit tests, but I understand that the PR is already big

Reply:
Ah is the answer that `D::update_external_component_access` must itself check for conflicts and possibly panic?

But in that case couldn't we just update `self.component_access` with the nested component accesses?
Where is the distinction between "current entity component access" and "nested entity component access" relevant?